core/src/main/java/org/sonarsource/sonarlint/core/telemetry/TelemetryLocalStorageManager.java

```diff
@@ -0,0 +1,128 @@
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.channels.FileLock;
+import java.nio.charset.StandardCharsets;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.StandardOpenOption;
+import java.time.OffsetDateTime;
+import java.util.Base64;
+import java.util.function.Consumer;
+import org.sonar.api.utils.log.Logger;
+import org.sonar.api.utils.log.Loggers;
+import org.sonarsource.sonarlint.core.client.api.util.SonarLintUtils;
+
+/**
+ * Serialize and deserialize telemetry data to persistent storage.
+ */
+class TelemetryLocalStorageManager {
+  private static final Logger LOG = Loggers.get(TelemetryLocalStorageManager.class);
+  private final Path path;
+
+  TelemetryLocalStorageManager(Path path) {
+    this.path = path;
+  }
+
+  private void updateAtomically(Consumer<TelemetryLocalStorage> updater) throws IOException {
+    Gson gson = createGson();
+    Files.createDirectories(path.getParent());
+    try (FileChannel fileChannel = FileChannel.open(path, StandardOpenOption.READ, StandardOpenOption.WRITE, StandardOpenOption.CREATE, StandardOpenOption.SYNC);
+      FileLock lock = fileChannel.lock()) {
+      TelemetryLocalStorage newData = read(gson, fileChannel);
+
+      updater.accept(newData);
+
+      fileChannel.truncate(0);
+
+      String newJson = gson.toJson(newData);
+      byte[] encoded = Base64.getEncoder().encode(newJson.getBytes(StandardCharsets.UTF_8));
+
+      fileChannel.write(ByteBuffer.wrap(encoded));
+    }
+  }
+
+  private TelemetryLocalStorage read(Gson gson, FileChannel fileChannel) throws IOException {
+    try {
+      if (fileChannel.size() == 0) {
+        return new TelemetryLocalStorage();
+      }
+      final ByteBuffer buf = ByteBuffer.allocate((int) fileChannel.size());
+      fileChannel.read(buf);
+      byte[] decoded = Base64.getDecoder().decode(buf.array());
+      String oldJson = new String(decoded, StandardCharsets.UTF_8);
+      TelemetryLocalStorage previousData = gson.fromJson(oldJson, TelemetryLocalStorage.class);
+
+      return TelemetryLocalStorage.validateAndMigrate(previousData);
+    } catch (Exception e) {
+      if (SonarLintUtils.isInternalDebugEnabled()) {
+        LOG.error("Error reading telemetry data", e);
+        throw new IllegalStateException(e);
+      }
+      return new TelemetryLocalStorage();
+    }
+  }
+
+  void tryUpdateAtomically(Consumer<TelemetryLocalStorage> updater) {
+    try {
+      updateAtomically(updater);
+    } catch (Exception e) {
+      if (SonarLintUtils.isInternalDebugEnabled()) {
+        LOG.error("Error updating telemetry data", e);
+        throw new IllegalStateException(e);
+      }
+    }
+  }
+
+  private static Gson createGson() {
+    return new GsonBuilder()
+      .registerTypeAdapter(OffsetDateTime.class, new OffsetDateTimeAdapter())
+      .create();
+  }
+
+  private TelemetryLocalStorage load() throws IOException {
```

Comment:
I am not sure I understand the difference between read and load. When to use one or the other ? Do we need both ? If so maybe we could come up with better names

Reply:
One method (`read`) is reading from the same FileChannel used later by write (in order to ensure atomicity). 
`load` is just reading without lock (mainly used to check the isEnabled flag).

Let me try to rename to make it more explicit.